File: Assets/Scripts/Weapon/Handgun.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

//手枪
public class Handgun : Gun
{
    protected override void Start()
    {
        base.Start();

        gunType = GunType.Short;

        DataInit();

    }

    protected override void Update()
    {
        base.Update();

        //Debug.Log("手枪开火间隔时长：" + fireTime);
    }

    //从Lua脚本获取数据
    protected override void DataInit()
    {
        maxMagazineSize = LuaManager.instance.handGunTab.Get<int>("maxMagazineSize");
        duration = LuaManager.instance.handGunTab.Get<float>("duration");
        maxAngleDeg = LuaManager.instance.handGunTab.Get<float>("maxAngleDeg");
    }


    //开火
    public override bool Fire(Player player, float moveSpeed, int fireCount)
    {
        if (currentMagazineSize > 0 && fireTime > LuaManager.instance.handGunTab.Get<float>("fireInterval"))
        {
            shootController.Fire(player, moveSpeed, fireCount, muzzleTransform, cam, maxAngleDeg);

            currentMagazineSize--;
            fireTime = 0;

            return true;
        }
        else
        {
            Debug.Log("弹夹为空||开火间隔");

            return false;
        }
    }

    //换弹
    public override void Reload(bool isReloading)
    {
        base.Reload(isReloading);
    }

    //获取枪支后坐力
    public override float GetGunRecoil(float elapsed)
    {
        return LuaManager.instance.getHandGunRecoil(elapsed);
    }

}

File: Assets/Lua/HandGun.lua
local HandGun = {
    maxMagazineSize = 7, --最大弹容量
    duration = 0.4, --枪口复位时间
    fireInterval = 0.1, --开火间隔时间
    maxAngleDeg = 4, --最大散射角度

    startAngle = 0, --当前后坐力角度
    targetOffset = 4 --后坐力威力
}

function HandGun.GetRecoil(elapsed)

    local t = elapsed / HandGun.duration

    if t < 0.3 then --上升
        local recoilT = t / 0.3
        local smoothT = recoilT * recoilT * (3 - 2 * recoilT)
        return CS.UnityEngine.Mathf.Lerp(HandGun.startAngle, HandGun.startAngle - HandGun.targetOffset, smoothT)
    else --下降
        local recoilT = (t - 0.3) / 0.7
        local smoothT = recoilT * recoilT * (3 - 2 * recoilT)
        return CS.UnityEngine.Mathf.Lerp(HandGun.startAngle - HandGun.targetOffset, HandGun.startAngle, smoothT)
    end

end

return HandGun